// SPDX-FileCopyrightText: 2018 - 2023 Deepin Technology Co., Ltd.
//
// SPDX-License-Identifier: LGPL-3.0-or-later

syntax = "proto3";

enum DeviceOS
{
    DEVICE_OS_OTHER = 0;
    DEVICE_OS_UOS = 1;
    DEVICE_OS_LINUX = 2;
    DEVICE_OS_WINDOWS = 3;
    DEVICE_OS_MACOS = 4;
    DEVICE_OS_ANDROID = 5;
}

enum Compositor
{
    COMPOSITOR_NONE = 0;
    COMPOSITOR_X11 = 1;
    COMPOSITOR_WAYLAND = 2;
}

message DeviceInfo
{
    string uuid = 1;
    string name = 2;            // 扫描发起者的设备名
    DeviceOS os = 3;            // 扫描发起者的操作系统
    Compositor compositor = 4;  // 扫描发起者的显示服务器类型
}

// UDP广播，端口号为 51595
message ScanRequest
{
    string key = 1;             // 固定值 "UOS-COOPERATION"
    DeviceInfo deviceInfo = 2;
    int32 port = 3;             // TCP 端口
}

// UDP，向 ScanRequest 来源的 IP:PORT 返回
message ScanResponse
{
    string key = 1;             // 固定值 "UOS-COOPERATION"
    DeviceInfo deviceInfo = 2;
    int32 port = 3;             // TCP 端口
}

// UDP，服务停止通知
message ServiceStoppedNotification
{
    string deviceUuid = 1;
}

// TCP，连接 ScanResponse 返回的端口
message PairRequest
{
    string key = 1;             // 固定值 "UOS-COOPERATION"
    DeviceInfo deviceInfo = 2;
}

// 配对时返回
message PairResponse
{
    string key = 1;             // 固定值 "UOS-COOPERATION"
    DeviceInfo deviceInfo = 2;
    bool agree = 3;             // 是否同意配对
}

// 从带外数据发送
message StopPairRequest
{
    bool immediately = 1;   // 是否立即断开,立即断开时不返回 response
}

message StopPairResponse
{
    string bye = 1; // 固定值 "UOS-BYE"
}
